dataclients/kubernetes/ingress.go

```diff
@@ -0,0 +1,775 @@
+			ic.ingress.Metadata.Name,
+			route.Id,
+			host+strings.Replace(path, "/", "_", -1),
+			extraIndex)
+		setPath(ic.pathMode, &route, path)
+		if n := countPathRoutes(&route); n <= 1 {
+			ic.addHostRoute(host, &route)
+			ic.redirect.updateHost(host)
+		} else {
+			log.Errorf("Failed to add route having %d path routes: %v", n, r)
+		}
+	}
+}
+
+// computeBackendWeights computes and sets the backend traffic weights on the
+// rule backends.
+// The traffic is calculated based on the following rules:
+//
+// * if no weight is defined for a backend it will get weight 0.
+// * if no weights are specified for all backends of a path, then traffic will
+//   be distributed equally.
+//
+// Each traffic weight is relative to the number of backends per path. If there
+// are multiple backends per path the weight will be relative to the number of
+// remaining backends for the path e.g. if the weight is specified as
+//
+//      backend-1: 0.2
+//      backend-2: 0.6
+//      backend-3: 0.2
+//
+// then the weight will be calculated to:
+//
+//      backend-1: 0.2
+//      backend-2: 0.75
+//      backend-3: 1.0
+//
+// where for a weight of 1.0 no Traffic predicate will be generated.
+func computeBackendWeights(backendWeights map[string]float64, rule *rule) {
+	type pathInfo struct {
+		sum          float64
+		lastActive   *backend
+		count        int
+		weightsCount int
+	}
+
+	// get backend weight sum and count of backends for all paths
+	pathInfos := make(map[string]*pathInfo)
+	for _, path := range rule.Http.Paths {
+		sc, ok := pathInfos[path.Path]
+		if !ok {
+			sc = &pathInfo{}
+			pathInfos[path.Path] = sc
+		}
+
+		if weight, ok := backendWeights[path.Backend.ServiceName]; ok {
+			sc.sum += weight
+			if weight > 0 {
+				sc.lastActive = path.Backend
+				sc.weightsCount++
+			}
+		} else {
+			sc.count++
+		}
+	}
+
+	// calculate traffic weight for each backend
+	for _, path := range rule.Http.Paths {
+		if sc, ok := pathInfos[path.Path]; ok {
+			if weight, ok := backendWeights[path.Backend.ServiceName]; ok {
+				// force a weight of 1.0 for the last backend with a non-zero weight to avoid rounding issues
+				if sc.lastActive == path.Backend {
+					path.Backend.Traffic = 1.0
+					continue
+				}
+
+				path.Backend.Traffic = weight / sc.sum
+				// subtract weight from the sum in order to
+				// give subsequent backends a higher relative
+				// weight.
+				sc.sum -= weight
+
+				// noops are required to make sure that routes are in order selected by
+				// routing tree
+				if sc.weightsCount > 2 {
+					path.Backend.noopCount = sc.weightsCount - 2
+				}
+				sc.weightsCount--
+			} else if sc.sum == 0 && sc.count > 0 {
+				path.Backend.Traffic = 1.0 / float64(sc.count)
+			}
+			// reduce count by one in order to give subsequent
+			// backends for the path a higher relative weight.
+			sc.count--
+		}
+	}
+}
+
+// TODO: default filters not applied to 'extra' routes from the custom route annotations. Is it on purpose?
```

Comment:
I think this should be fixed, if it was done like that. I remember, that I changed it to apply everything from the original ingress including filters and predicates, such that custom routes are bound to the ingress and can not increase the scope, but always have to reduce the scope.

Reply:
https://github.com/zalando/skipper/issues/1287